h2o-core/src/main/java/water/rapids/ast/prims/string/AstReplaceAll.java

```diff
@@ -65,16 +66,29 @@ public Val apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
   }
 
   private Vec replaceAllCategoricalCol(Vec vec, String pattern, String replacement, boolean ignoreCase) {
+    final Pattern compiledPattern = Pattern.compile(pattern); // Compile the pattern once before replacement
     String[] doms = vec.domain().clone();
-    for (int i = 0; i < doms.length; ++i)
+    Set<String> newDomainSet = new HashSet<>(); // The pattern might create multiple domains with the same name
+    for (int i = 0; i < doms.length; ++i) {
       doms[i] = ignoreCase
-          ? doms[i].toLowerCase(Locale.ENGLISH).replaceAll(pattern, replacement)
-          : doms[i].replaceAll(pattern, replacement);
+              ? compiledPattern.matcher(doms[i].toLowerCase(Locale.ENGLISH)).replaceAll(replacement)
+              : compiledPattern.matcher(doms[i]).replaceAll(replacement);
+      newDomainSet.add(doms[i]);
+    }
 
-    return vec.makeCopy(doms);
+    if (newDomainSet.size() == doms.length) {
+      // Avoid remapping if cardinality is the same
+      newDomainSet = null;
+      return vec.makeCopy(doms);
+    } else {
+      newDomainSet = null;
+      final Vec copy = vec.makeCopy();
```

Comment:
Hmm, this creates a deep copy of the data just to overwrite it with something else.

I think it would be nicer if the RemapDomainTask simply produced a new vector (instead of mutating an existing duplicate version).

Reply:
Since the old logic was mutating state of a deep copy of the original vec, this approach made sense. 

What doesn't make sense is to create a deep copy & then mutate it, making this approach not justifiable as well.

Reply:
Correct. Removed. Addressed this problem. The MR task itself now makes the task distributed uses NewChunk to pump data into a newly created Vec. No more deep copying and then "iterating" again, resulting in O(2n) on this level.